File: java/org.kevoree.library.java.javaNode/src/main/java/org/kevoree/library/java/planning/KevoreeScheduler.java
package org.kevoree.library.java.planning;

import org.kevoree.api.adaptation.AdaptationModel;
import org.kevoree.api.adaptation.AdaptationPrimitive;
import org.kevoree.api.adaptation.AdaptationType;
import org.kevoree.api.adaptation.Step;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;


/**
 * Created by IntelliJ IDEA.
 * User: duke
 * Date: 21/09/11
 * Time: 17:54
 */

public abstract class KevoreeScheduler {

    public AdaptationModel schedule(AdaptationModel adaptionModel) {
        if (!adaptionModel.getAdaptations().isEmpty()) {
            HashMap<String, List<AdaptationPrimitive>> classedAdaptations = classify(adaptionModel.getAdaptations());
            adaptionModel.setOrderedPrimitiveSet(createStep(classedAdaptations.get(AdaptationType.AddDeployUnit.name()),AdaptationType.AddDeployUnit));
            Step firstStep = adaptionModel.getOrderedPrimitiveSet();
            firstStep.next(createStep(classedAdaptations.get(AdaptationType.StopInstance.name()),AdaptationType.StopInstance))
                    .next(createStep(classedAdaptations.get(AdaptationType.RemoveBinding.name()),AdaptationType.RemoveBinding))
                    .next(createStep(classedAdaptations.get(AdaptationType.RemoveInstance.name()),AdaptationType.RemoveInstance))
                    .next(createStep(classedAdaptations.get(AdaptationType.RemoveDeployUnit.name()),AdaptationType.RemoveDeployUnit))
                    .next(createStep(classedAdaptations.get(AdaptationType.AddInstance.name()),AdaptationType.AddInstance))
                    .next(createStep(classedAdaptations.get(AdaptationType.AddBinding.name()),AdaptationType.AddBinding))
                    .next(createStep(classedAdaptations.get(AdaptationType.UpdateDictionaryInstance.name()),AdaptationType.UpdateDictionaryInstance))
                    .next(createStep(classedAdaptations.get(AdaptationType.UpdateCallMethod.name()),AdaptationType.UpdateCallMethod))
                    .next(createStep(classedAdaptations.get(AdaptationType.StartInstance.name()),AdaptationType.StartInstance));
        } else {
            adaptionModel.setOrderedPrimitiveSet(null);
        }
        return adaptionModel;
    }

    private HashMap<String, List<AdaptationPrimitive>> classify(List<AdaptationPrimitive> inputs) {
        HashMap<String, List<AdaptationPrimitive>> result = new HashMap<String, List<AdaptationPrimitive>>();
        for (AdaptationPrimitive adapt : inputs) {
            List<AdaptationPrimitive> l;
            if (!result.containsKey(adapt.getPrimitiveType())) {
                l = new ArrayList<AdaptationPrimitive>();
                result.put(adapt.getPrimitiveType(), l);
            } else {
                l = result.get(adapt.getPrimitiveType());
            }
            l.add(adapt);
        }
        return result;
    }

    public Step createStep(List<AdaptationPrimitive> commands, AdaptationType type) {
        Step step = new Step();
        step.setAdaptationType(type);
        if (commands != null) {
            step.getAdaptations().addAll(commands);
        }
        return step;
    }
}

File: java/org.kevoree.library.java.javaNode/src/main/java/org/kevoree/library/java/planning/KevoreeScheduler.kt
package org.kevoree.library.java.planning

import java.util.HashMap
import java.util.ArrayList
import org.kevoree.factory.KevoreeFactory
import org.kevoree.factory.DefaultKevoreeFactory
import org.kevoree.api.adaptation.AdaptationModel
import org.kevoree.api.adaptation.Step
import org.kevoree.api.adaptation.AdaptationPrimitive
import org.kevoree.api.adaptation.SequentialStep


/**
 * Created by IntelliJ IDEA.
 * User: duke
 * Date: 21/09/11
 * Time: 17:54
 */

trait KevoreeScheduler {

    var adaptationModelFactory: KevoreeFactory

    open fun schedule(adaptionModel: AdaptationModel, nodeName: String): AdaptationModel {
        if (!adaptionModel.adaptations.isEmpty()) {
            adaptationModelFactory = DefaultKevoreeFactory()
            val classedAdaptations = classify(adaptionModel.adaptations)
            adaptionModel.orderedPrimitiveSet = createStep(classedAdaptations.get(JavaPrimitive.AddDeployUnit.name()))
            var currentStep : Step? = adaptionModel.orderedPrimitiveSet
            currentStep!!.nextStep = createStep(classedAdaptations.get(JavaPrimitive.LinkDeployUnit.name()))
            currentStep = currentStep!!.nextStep
            currentStep!!.nextStep = createStep(classedAdaptations.get(JavaPrimitive.AddInstance.name()))
            currentStep = currentStep!!.nextStep
            currentStep!!.nextStep = createStep(classedAdaptations.get(JavaPrimitive.StopInstance.name()))
            currentStep = currentStep!!.nextStep
            currentStep!!.nextStep = createStep(classedAdaptations.get(JavaPrimitive.RemoveBinding.name()))
            currentStep = currentStep!!.nextStep
            currentStep!!.nextStep = createStep(classedAdaptations.get(JavaPrimitive.UpgradeInstance.name()))
            currentStep = currentStep!!.nextStep
            currentStep!!.nextStep = createStep(classedAdaptations.get(JavaPrimitive.RemoveInstance.name()))
            currentStep = currentStep!!.nextStep
            currentStep!!.nextStep = createStep(classedAdaptations.get(JavaPrimitive.AddBinding.name()))
            currentStep = currentStep!!.nextStep
            currentStep!!.nextStep = createStep(classedAdaptations.get(JavaPrimitive.UpdateDictionaryInstance.name()))
            currentStep = currentStep!!.nextStep
            currentStep!!.nextStep = createStep(classedAdaptations.get(JavaPrimitive.UpdateCallMethod.name()))
            currentStep = currentStep!!.nextStep
            currentStep!!.nextStep = createStep(classedAdaptations.get(JavaPrimitive.StartInstance.name()))
            currentStep = currentStep!!.nextStep
            currentStep!!.nextStep = createStep(classedAdaptations.get(JavaPrimitive.RemoveDeployUnit.name()))
        } else {
            adaptionModel.orderedPrimitiveSet = null
        }
        return adaptionModel
    }

    private fun classify(inputs: List<AdaptationPrimitive>): HashMap<String, MutableList<AdaptationPrimitive>> {
        var result = HashMap<String, MutableList<AdaptationPrimitive>>()
        for(adapt in inputs){
            var l: MutableList<AdaptationPrimitive>? = null
            if(!result.containsKey(adapt.primitiveType)){
                l = ArrayList<AdaptationPrimitive>()
                result.put(adapt.primitiveType!!, l!!)
            } else {
                l = result.get(adapt.primitiveType!!)
            }
            l!!.add(adapt)

        }
        return result
    }

    public fun createStep(commands: MutableList<AdaptationPrimitive>?): Step {
      //  var currentSteps = adaptationModelFactory.createParallelStep()
        var currentSteps = SequentialStep()
        if(commands != null){
            currentSteps.adaptations.addAll(commands)
        }
        return currentSteps
    }


}